docs/spec/relayer/Relayer.md

```diff
@@ -0,0 +1,240 @@
+# Relayer Specification
+
+Relayers are processes that provide connection layer of the IBC protocol. In the IBC protocol, on chain
+modules do not have a way of directly sending a message to each other; this is the responsibility of relayer
+processes. Modules signal its intention to send a message by writing data in its data store at the
+defined location, and make those data (with corresponding proofs) available to external parties.
+Relayer processes read (we say also scan) the state of each chain, construct appropriate IBC datagrams,
+verify the corresponding proofs and submit valid datagrams to destination chain.   
+We assume existence of multiple relayers, where some relayers could be faulty (behave arbitrarily),
+but there is always at least a single correct relayer. We don't make assumptions on the maximum number of 
+faulty relayers.
+
+For the purpose of this specification we assume existence of two on chain modules A and B, that executes
+IBC protocol. We say that a module A (or B) sends an IBC datagram m to a module B (or A) when a correct
+relayer can construct valid datagram m by scanning the state of the chain A. We say that a module A receives
+an IBC datagram m, when m was processed by the module A on chain. We assume that modules
+are correct.    
+
+Correct relayers need to ensure the following properties:
+
+**[ICS18-Delivery]**: If a module A sends an IBC datagram m to a module B, then m is
```

Comment:
I wonder if it would help to be more accurate here. I see two issues: 

1. The notion of "send" for a module is not well-defined. Instead, I guess what happens is: a module A emits IBC events, and some of these events indicate state transitions that are relevant for another module B.

2. Perhaps we should refer to chains here, not modules. The function signature `func handleEvent(ev, chainA, chainB)` refers to chains, and there isn't reference to modules anywhere except in these properties. Indeed, relaying is done at chain-level. Relayer granularity seems impossible at module-level.

Proposal for refining the properties based on these ideas:

**[ICS18-Delivery]**: If any module on `chainA` emits an IBC event `m` relevant for some module on `chainB`, then `chainB` eventually receives an IBC datagram matching event `m`.

**[ICS18-Validity]**: If some module on `chainB` receives an IBC datagram for an event `m` from a module on `chainA`, then a module on `chainA` emitted an event signalling event `m` to a module on `chainB`.

This is not perfect because we need to define what it means for an event to be "relevant" for a module. We can clarify this with specific examples.